src/execution_plan/execution_plan.c

```diff
@@ -8,695 +8,804 @@
-                    Vector_Push(ops, op);
-                }
-            }
-            Vector_Clear(traversals);
-        }
-        Vector_Free(traversals);
+    // Build traversal operations for every connected component in the QueryGraph
+    if (AST_ContainsClause(ast, CYPHER_AST_MATCH) || AST_ContainsClause(ast, CYPHER_AST_MERGE)) {
+        _ExecutionPlanSegment_ProcessQueryGraph(segment, qg, filter_tree, resolved, ops);
     }
 
-    if(ast->unwindNode) {
-        OpBase *opUnwind = NewUnwindOp(ast);
+    // Set root operation
+    const cypher_astnode_t *unwind_clause = AST_GetClause(ast, CYPHER_AST_UNWIND);
+    if(unwind_clause) {
+        AST_UnwindContext unwind_ast_ctx = AST_PrepareUnwindOp(unwind_clause, record_map);
+
+        OpBase *opUnwind = NewUnwindOp(unwind_ast_ctx.record_idx, unwind_ast_ctx.exps);
         Vector_Push(ops, opUnwind);
         _UpdateResolvedVariables(resolved, opUnwind);
     }
 
-    /* Set root operation */
-    if(ast->createNode) {
-        BuildQueryGraph(gc, q, ast->createNode->graphEntities);
-        OpBase *opCreate = NewCreateOp(ctx, ast, q, execution_plan->result_set);
-
+    bool create_clause = AST_ContainsClause(ast, CYPHER_AST_CREATE);
+    if(create_clause) {
+        QueryGraph_AddCreateClauses(gc, ast, qg);
+        AST_CreateContext create_ast_ctx = AST_PrepareCreateOp(record_map, ast, qg);
+        OpBase *opCreate = NewCreateOp(&result_set->stats,
+                                       create_ast_ctx.nodes_to_create,
+                                       create_ast_ctx.edges_to_create);
         Vector_Push(ops, opCreate);
         _UpdateResolvedVariables(resolved, opCreate);
     }
 
-    if(ast->mergeNode) {
-        OpBase *opMerge = NewMergeOp(ast, execution_plan->result_set);
+    const cypher_astnode_t *merge_clause = AST_GetClause(ast, CYPHER_AST_MERGE);
+    if(merge_clause) {
+        // A merge clause provides a single path that must exist or be created.
+        // As with paths in a MATCH query, build the appropriate traversal operations
+        // and append them to the set of ops.
+        AST_MergeContext merge_ast_ctx = AST_PrepareMergeOp(record_map, ast, merge_clause, qg);
+
+        // Append a merge operation
+        OpBase *opMerge = NewMergeOp(&result_set->stats,
+                                     merge_ast_ctx.nodes_to_merge,
+                                     merge_ast_ctx.edges_to_merge);
         Vector_Push(ops, opMerge);
         _UpdateResolvedVariables(resolved, opMerge);
     }
 
-    if(ast->deleteNode) {
-        OpBase *opDelete = NewDeleteOp(ast->deleteNode, q, execution_plan->result_set, ast);
+    const cypher_astnode_t *delete_clause = AST_GetClause(ast, CYPHER_AST_DELETE);
+    if(delete_clause) {
+        uint *nodes_ref;
+        uint *edges_ref;
+        AST_PrepareDeleteOp(delete_clause, qg, record_map, &nodes_ref, &edges_ref);
+        ResultSetStatistics *stats = (result_set) ? &result_set->stats : NULL;
```

Comment:
in NewMergeOp we didn't check to see if result_set is NULL, should we?

Reply:
We should, thanks!